src/DocumentDB.ChangeFeedProcessor/IUnhealthyHandlingStrategy.cs

```diff
@@ -0,0 +1,24 @@
+﻿//----------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  Licensed under the MIT license.
+//----------------------------------------------------------------
+
+namespace Microsoft.Azure.Documents.ChangeFeedProcessor
+{
+    using System;
+    using System.Threading.Tasks;
+
+    internal interface IUnhealthyHandlingStrategy
+    {
+        Task HandleAsync();
+    }
+
+    internal class FailFastUnhealthyHandlingStrategy : IUnhealthyHandlingStrategy
+    {
+        public Task HandleAsync()
+        {
+            Environment.FailFast("Unhealthy instance detected.");
```

Comment:
I like the idea in general but I don't think we can do this by default. Otherwise we may get lots of noise like "CFP is crashing for me, please help!!".

Here's what I think:
- no crash by default
- duration of "how long leases cannot be acquired" should be user-provided with some reasonable default. I don't think lease expiration * 5 is good, should be longer, maybe 10 minutes or so.
- the action should be user-provided and by default no action (should trace error though). The strategy can be provided as something like builder.WithHealthStrategy or .WithHealthMonitor. When we detect potential unhealthy condition, we should call user-provided event passing type of unhealthy condition and some (optional?) description. The user's event handler decides what to do: trace, make internal action, stop the processor, FailFast, etc.

Reply:
YEAH!

Default is SilentUnhealthinessHandlingStrategy. So I set it to Max(15mins or 15x lease expiration interval).

Reply:
Can we make more generic, at least that unhealthy condition can be set from outside?

Reply:
Currently this is generic name but only handles one lease issue. So still, can we have a more generic strategy? So that the decision on how many times or how long time would come from the user? We can use it for other cases, like 500s/exceptions coming from the server in the change feed query loop, split issues, etc. Something like this:

enum HealthEventLevel { Error, Warning /* others can be added later */}
enum HealthEventType { UpdateLeaseLeaseError, ServerError,... }
interface IHealthMonitor {
  void HealthEvent(HealthEventLevel level, healthEventType eventType, string description, /* optional*/ Exception exception);
}
ChangeFeedProcessorBuilder.WithHealthMonotor(...)

What do you think?